// eslint-config-ali/rules/ali/base.js
module.exports = {
  rules: {
    /**
     * best-practices
     */
    // 不允许修改函数的入参
    // airbnb推荐值：[2, { 'props': true }] disallow reassignment of function parameters and parameter object manipulation
    'no-param-reassign': 0,
    // 不允许不执行的表达式（需要函数调用或赋值） disallow usage of expressions in statement position
    // airbnb推荐值：2
    'no-unused-expressions': [2, { allowShortCircuit: true, allowTernary: true }],
    // for in循环中必须使用hasOwnProperty.call做判断
    // airbnb推荐值：2
    'guard-for-in': 0,
    // disallow the use of leading or trailing decimal points in numeric literals
    // airbnb推荐值：2
    'no-floating-decimal': 1,
    // disallow certain syntax forms
    // airbnb推荐值：'ForInStatement'也不允许
    'no-restricted-syntax': [
      2,
      'DebuggerStatement',
      'LabeledStatement',
      'WithStatement',
    ],
    // require return statements to either always or never specify values
    // airbnb推荐值：2，置为关闭
    'consistent-return': 0,
    // disallow else after a return in an if
    // airbnb推荐值：2，意义不大，先改为warn
    'no-else-return': 1,
    // requires to declare all vars on top of their containing scope
    // airbnb推荐值：2
    'vars-on-top': 1,
    // require the use of === and !==
    // airbnb推荐值：['error', 'allow-null']，降低报错级别
    eqeqeq: [1, 'allow-null'],
    // require use of the second argument for parseInt()
    // airbnb推荐值：'error'
    radix: 'warn',

    /**
     * style
     */
    // 尽量使用驼峰命名
    // airbnb推荐值：[2, { 'properties': 'never' }]
    camelcase: [1, { properties: 'never' }],
    // disallow dangling underscores in identifiers
    // airbnb推荐值：[2, { 'allowAfterThis': false }]
    'no-underscore-dangle': 0,
    // enforces new line after each method call in the chain to make it
    // more readable and easy to maintain
    // airbnb推荐值：[2, { 'ignoreChainWithDepth': 3 }]
    'newline-per-chained-call': 0,
    // require function expressions to have a name
    // airbnb推荐值：1，我们的老代码里有很多用匿名函数的地方，所以先去掉这条规则
    'func-names': 0,
    // specify the maximum length of a line in your program
    // airbnb推荐值：[2, 100, ...]，先设为1并且扩大长度到200
    'max-len': [1, 200, 2, {
      ignoreUrls: true,
      ignoreComments: false
    }],
    // require a capital letter for constructors
    // airbnb推荐值：[2, { 'newIsCap': true }]，为了js-to-java包改一下
    'new-cap': [2, { newIsCap: true, properties: false }]
  },
  env: {
    // 默认打开常用环境
    browser: true,
    node: true,
    jquery: true
  }
};
